pacmacs.el

```diff
@@ -322,46 +323,72 @@
-              (dolist (candidate-way candidate-ways)
-                (pacmacs--track-point candidate-way p))
+                    (cl-remove-if #'pacmacs--filter-candidates
+                                  possible-ways))
+                   (candidate-distance (1+ (pacmacs--cell-wrapped-get
+                                            pacmacs--track-board
+                                            current-row current-column))))
+              (-each candidate-ways
+                (-lambda ((candidate-row . candidate-column))
+                  (pacmacs--cell-wrapped-set pacmacs--track-board
+                                             candidate-row
+                                             candidate-column
+                                             candidate-distance)))
               (setq next-wave
                     (append next-wave candidate-ways))))
           (setq wave next-wave))))))
 
-(defun pacmacs--track-object (game-object)
+(defun pacmacs--track-object-direction (game-object sort-comparator)
   (plist-bind ((row :row)
                (column :column))
       game-object
-    (let ((direction (pacmacs--cell-wrapped-get pacmacs--track-board row column)))
-      (pacmacs--switch-direction game-object direction))))
+    (let* ((candidate-ways (cons
+                            (cons row column)
+                            (-remove (-lambda ((candidate-row . candidate-column))
+                                       (pacmacs--wall-at-p candidate-row candidate-column))
+                                     (pacmacs--possible-side-ways row column))))
+           (candidate-distances (-map (-lambda ((candidate-row . candidate-column))
+                                        (pacmacs--cell-wrapped-get pacmacs--track-board
+                                                                   candidate-row
+                                                                   candidate-column))
+                                      candidate-ways))
+           (next-tile (->> (-zip candidate-distances candidate-ways)
+                           (-sort sort-comparator)
+                           (cdar))))
+      (when next-tile
+        (->> (pacmacs--vector- next-tile (cons row column))
+             (pacmacs--direction-name))))))
+
+(defun pacmacs--track-object (game-object)
+  (-when-let (direction (pacmacs--track-object-direction
+                         game-object
+                         (-lambda ((distance-1 . _) (distance-2 . _))
+                           (< distance-1 distance-2))))
+    (pacmacs--switch-direction game-object direction)))
+
+(defun pacmacs--back-track-object (game-object)
+  (-when-let (direction (pacmacs--track-object-direction
+                         game-object
+                         (-lambda ((distance-1 . _) (distance-2 . _))
+                           (> distance-1 distance-2))))
+    (pacmacs--switch-direction game-object direction)))
```

Comment:
Looks like `pacmacs--track-object` and `pacmacs--back-track-object` have a duplicate code. Just a reminder to myself to create a general function.